src/main/java/com/spotify/autoscaler/api/KCCAutoscalerController.java

```diff
@@ -59,66 +59,86 @@ public KCCAutoscalerController(final Database database) {
-        Arrays.stream(autoscalerConfig.getSpec().getCluster())
-            .collect(
-                Collectors.toMap(
-                    BigtableAutoscaler.Spec.Cluster::getClusterId, Function.identity()));
+    return reconcileResponse(projectId, instanceId, clustersDesiredState, forObj);
+  }
 
-    targetClusters.forEach(
-        (clusterId, clusterSpec) ->
+  private void reconcileClustersDesiredState(
+      final String projectId,
+      final String instanceId,
+      final List<BigtableAutoscaler.Spec.Cluster> clustersDesiredState) {
+    clustersDesiredState.forEach(
+        cluster ->
             database.reconcileBigtableCluster(
                 new BigtableClusterBuilder()
                     .projectId(projectId)
                     .instanceId(instanceId)
-                    .clusterId(clusterId)
-                    .minNodes(clusterSpec.getMinNodes())
-                    .maxNodes(clusterSpec.getMaxNodes())
-                    .cpuTarget(clusterSpec.getCpuTarget())
+                    .clusterId(cluster.getClusterId())
+                    .minNodes(cluster.getMinNodes())
+                    .maxNodes(cluster.getMaxNodes())
+                    .cpuTarget(cluster.getCpuTarget())
                     .enabled(true)
                     .build()));
+    deleteAbsentClusters(projectId, instanceId, clustersDesiredState);
+  }
 
-    int deletedClustersCount =
-        database.deleteBigtableClustersExcept(projectId, instanceId, targetClusters.keySet());
+  private void deleteAbsentClusters(
+      final String projectId,
+      final String instanceId,
+      final List<BigtableAutoscaler.Spec.Cluster> clustersDesiredState) {
+    final int deletedClustersCount =
+        database.deleteBigtableClustersExcept(
+            projectId, instanceId, getClusterIds(clustersDesiredState));
     if (deletedClustersCount > 0) {
       LOGGER.info(
           "Project {}, instance {}: {} clusters deleted",
           projectId,
           instanceId,
           deletedClustersCount);
     }
+  }
 
-    final List<BigtableCluster> clusters = database.getBigtableClusters(projectId, instanceId);
-    forObj.status(
-        gson.toJsonTree(new ReconcileResponse.Status(clusters), ReconcileResponse.Status.class));
+  private Set<String> getClusterIds(
+      final List<BigtableAutoscaler.Spec.Cluster> clustersDesiredState) {
+    return clustersDesiredState
+        .stream()
+        .map(BigtableAutoscaler.Spec.Cluster::getClusterId)
+        .collect(Collectors.toSet());
+  }
 
+  private String reconcileResponse(
+      final String projectId,
+      final String instanceId,
+      final List<BigtableAutoscaler.Spec.Cluster> clustersDesiredState,
+      final ReconcileResponse.EnsureResources.ForObject forObj) {
+    final List<BigtableCluster> clustersFromDB =
+        database.getBigtableClusters(projectId, instanceId);
+    if (clustersDesiredState.size() != clustersFromDB.size()) {
+      throw new IllegalStateException(
```

Comment:
This would mean that the reconcile would stop working if the user start messing up with the other API?

Reply:
Nevermind, I think it's fine!